### PR TITLE
TokenRegistry updates (Issue #174, #155, #173)

### PR DESCRIPTION
**Pull request by Brechtpd**
See issue #174, #155, #173 
Still needs testing.

### CODE REVIEW

**Comment by dong77** [comment]
I think we need to change `unregisterToken` to take O(1) time first, as it will probably affect the overall storage.

File: contracts/TokenRegistry.sol [hunk 1]
```diff
@@ -40,7 +52,7 @@ contract TokenRegistry is Claimable {
         require(!isTokenRegisteredBySymbol(_symbol));
```
Comment (dong77) [inline contracts/TokenRegistry.sol]: can we also check `_symbol` is not "" by using `bytes(_symbol).length > 0`

File: contracts/TokenRegistry.sol [hunk 2]
```diff
@@ -28,10 +28,22 @@ contract TokenRegistry is Claimable {
...
+    
+    struct TokenInfo {
+        uint index;             // Index + 1 of the address in the tokens array, 0 means unregistered.
+        uint8 tokenStandard;    // ERC20 or ERC223
+        string symbol;          // Symbol of the token
+    }
+    
 
+    // TODO: need to add token standard
     function registerToken(address _token, string _symbol)
```
Comment (dong77) [inline contracts/TokenRegistry.sol]: renaming _token to _addr should be a good idea.

File: contracts/TokenRegistry.sol [hunk 3]
```diff
@@ -51,14 +63,25 @@ contract TokenRegistry is Claimable {
...
         delete tokenMap[_token];
-        for (uint i = 0; i < tokens.length; i++) {
-            if (tokens[i] == _token) {
-                tokens[i] = tokens[tokens.length - 1];
-                tokens.length --;
-                break;
-            }
+        
+        // Stored index in TokenInfo is index+1
+        index--;
```
Comment (dong77) [inline contracts/TokenRegistry.sol]: I would suggest to remove this line (ie `index--`), and change:
```
tokens[index] = lastToken;
tokenMap[lastToken].index = index + 1;
```
to:
```
tokens[index - 1] = lastToken;
tokenMap[lastToken].index = index ;
```

File: contracts/TokenRegistry.sol [hunk 4]
```diff
@@ -97,4 +130,31 @@ contract TokenRegistry is Claimable {
...
+            return;
+        }
+        
+        uint endIdx = startIdx + count;
+        if (endIdx > numTokens) {
+            endIdx = numTokens;
+        }
+
+        uint subListLength = endIdx - startIdx;
+        if (subListLength == 0) {
```
Comment (dong77) [inline contracts/TokenRegistry.sol]:
```
if (endIdx == strastIdx) {
    return;
}
``` 
then

```
tokensSubList = new address[](endIdx - startIdx);

```

?

**Comment by dong77** [comment]
If you would add tests to cover the  core logics. I'll send you more LRC :)

**Comment by Brechtpd** [comment]
Great suggestions. Didn't know version1.0 was the base for this, should be fine now.
Will write the tests tomorrow!
ETH address: 0xa3e90bAA36fABc9D110b4bEa909377853e6D8b62

**Comment by dong77** [comment]
https://etherscan.io/tx/0xb547786a5f7702a8eef9d82e276b483d233ff2bedc4bef2fc56ef99bfee555e7 @Brechtpd

**Comment by Brechtpd** [comment]
Thanks!

**Pull request merged**